File: app/src/main/java/com/ababilx/routinescrapper/ui/student/components/DateStrip.kt
package com.ababilx.routinescrapper.ui.student.components

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.lazy.LazyRow
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.unit.dp
import com.ababilx.routinescrapper.domain.model.RoutineDay
import com.ababilx.routinescrapper.ui.theme.Ink
import com.ababilx.routinescrapper.ui.theme.Line
import com.ababilx.routinescrapper.ui.theme.OnInk
import com.ababilx.routinescrapper.ui.theme.RdiuTypography
import com.ababilx.routinescrapper.ui.theme.TextMuted
import java.util.Calendar

data class DayChip(
    val day: RoutineDay,
    val date: Int,
)

@Composable
fun DateStrip(
    selected: RoutineDay,
    onSelect: (RoutineDay) -> Unit,
    modifier: Modifier = Modifier,
) {
    LazyRow(
        modifier = modifier,
        horizontalArrangement = Arrangement.spacedBy(12.dp),
    ) {
        items(weekChips(), key = { it.day }) { chip ->
            val active = chip.day == selected
            Column(
                modifier = Modifier
                    .clip(RoundedCornerShape(22.dp))
                    .clickable { onSelect(chip.day) }
                    .padding(horizontal = 4.dp, vertical = 2.dp),
                horizontalAlignment = Alignment.CenterHorizontally,
                verticalArrangement = Arrangement.spacedBy(8.dp),
            ) {
                Box(
                    modifier = Modifier
                        .size(48.dp)
                        .clip(CircleShape)
                        .background(if (active) Ink else Line.copy(alpha = 0.45f)),
                    contentAlignment = Alignment.Center,
                ) {
                    Text(
                        "${chip.date}",
                        style = RdiuTypography.titleMedium,
                        color = if (active) OnInk else TextMuted,
                    )
                }
                Text(
                    chip.day.shortLabel.uppercase(),
                    style = RdiuTypography.labelSmall,
                    color = if (active) Ink else TextMuted,
                )
            }
        }
    }
}

fun weekChips(): List<DayChip> {
    val start = Calendar.getInstance()
    val daysFromSaturday =
        (start.get(Calendar.DAY_OF_WEEK) - Calendar.SATURDAY + 7) % 7
    start.add(Calendar.DAY_OF_YEAR, -daysFromSaturday)
    return RoutineDay.entries.mapIndexed { index, day ->
        val dayCal = start.clone() as Calendar
        dayCal.add(Calendar.DAY_OF_YEAR, index)
        DayChip(day, dayCal.get(Calendar.DAY_OF_MONTH))
    }
}

File: lib/ui/student/components/date_strip.dart
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

import '../../../domain/model/routine_day.dart';
import '../../theme/app_colors.dart';

class DayChip {
  const DayChip(this.day, this.date, this.isToday);

  final RoutineDay day;
  final int date;
  final bool isToday;
}

class DateStrip extends StatefulWidget {
  const DateStrip({
    super.key,
    required this.selected,
    required this.today,
    required this.onSelect,
  });

  final RoutineDay selected;
  final RoutineDay today;
  final ValueChanged<RoutineDay> onSelect;

  @override
  State<DateStrip> createState() => _DateStripState();
}

class _DateStripState extends State<DateStrip> {
  late final ScrollController _scroll;

  @override
  void initState() {
    super.initState();
    _scroll = ScrollController();
    WidgetsBinding.instance.addPostFrameCallback((_) => _scrollToToday());
  }

  @override
  void didUpdateWidget(DateStrip oldWidget) {
    super.didUpdateWidget(oldWidget);
    if (oldWidget.today != widget.today) {
      _scrollToToday();
    }
  }

  @override
  void dispose() {
    _scroll.dispose();
    super.dispose();
  }

  void _scrollToToday() {
    final chips = weekChips(widget.today);
    final index = chips
        .indexWhere((chip) => chip.isToday)
        .clamp(0, chips.length);
    if (!_scroll.hasClients) return;
    const itemExtent = 60.0;
    _scroll.animateTo(
      (index * itemExtent).clamp(0, _scroll.position.maxScrollExtent),
      duration: const Duration(milliseconds: 280),
      curve: Curves.easeOut,
    );
  }

  @override
  Widget build(BuildContext context) {
    final chips = weekChips(widget.today);
    return SizedBox(
      height: 82,
      child: ListView.separated(
        controller: _scroll,
        scrollDirection: Axis.horizontal,
        itemCount: chips.length,
        separatorBuilder: (_, _) => const SizedBox(width: 12),
        itemBuilder: (context, index) {
          final chip = chips[index];
          final active = chip.day == widget.selected;
          final circleColor = active
              ? ink
              : chip.isToday
              ? mint
              : line.withValues(alpha: 0.45);
          return GestureDetector(
            onTap: () {
              HapticFeedback.selectionClick();
              widget.onSelect(chip.day);
            },
            child: Padding(
              padding: const EdgeInsets.symmetric(horizontal: 4, vertical: 2),
              child: Column(
                children: [
                  AnimatedContainer(
                    duration: const Duration(milliseconds: 220),
                    width: 48,
                    height: 48,
                    decoration: BoxDecoration(
                      color: circleColor,
                      shape: BoxShape.circle,
                      border: chip.isToday && !active
                          ? Border.all(
                              color: ink.withValues(alpha: 0.25),
                              width: 2,
                            )
                          : null,
                    ),
                    alignment: Alignment.center,
                    child: Text(
                      '${chip.date}',
                      style: Theme.of(context).textTheme.titleMedium?.copyWith(
                        color: active ? onInk : ink,
                      ),
                    ),
                  ),
                  const SizedBox(height: 6),
                  Text(
                    chip.isToday ? 'আজ' : chip.day.shortLabel.toUpperCase(),
                    style: Theme.of(context).textTheme.labelSmall?.copyWith(
                      color: active || chip.isToday ? ink : textMuted,
                    ),
                  ),
                ],
              ),
            ),
          );
        },
      ),
    );
  }
}

List<DayChip> weekChips([
  RoutineDay today = RoutineDay.saturday,
  DateTime? now,
]) {
  final start = now ?? DateTime.now();
  final daysFromSaturday = _javaDayOfWeek(start) % 7;
  final saturday = DateTime(
    start.year,
    start.month,
    start.day - daysFromSaturday,
  );
  return [
    for (var i = 0; i < RoutineDay.values.length; i++)
      DayChip(
        RoutineDay.values[i],
        DateTime(saturday.year, saturday.month, saturday.day + i).day,
        RoutineDay.values[i] == today,
      ),
  ];
}

int _javaDayOfWeek(DateTime date) {
  return date.weekday == DateTime.sunday ? 1 : date.weekday + 1;
}
